server/src/hover.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) ZeroC, Inc.
+
+use slicec::{
+    compilation_state::CompilationState,
+    grammar::{
+        Class, CustomType, Enum, Enumerator, Exception, Field, Interface, Module, Operation,
+        Parameter, Primitive, Struct, TypeAlias, TypeRef, TypeRefDefinition, Types,
+    },
+    slice_file::{Location, SliceFile, Span},
+    visitor::Visitor,
+};
+use tower_lsp::lsp_types::{Position, Url};
+
+pub fn get_hover_info(state: &CompilationState, uri: Url, position: Position) -> Option<String> {
+    // Attempt to convert the URL to a file path and then to a string
+    let file_path = uri.to_file_path().ok()?.to_str()?.to_owned();
+
+    // Attempt to retrieve the file from the state
+    let file = state.files.get(&file_path)?;
+
+    // Convert position to row and column 1 based
+    let col = (position.character + 1) as usize;
+    let row = (position.line + 1) as usize;
+
+    let mut visitor = HoverVisitor::new(slicec::slice_file::Location { row, col });
+    file.visit_with(&mut visitor);
+
+    visitor.found_message
+}
+
+struct HoverVisitor {
+    pub search_location: Location,
+    pub found_message: Option<String>,
+}
+
+impl HoverVisitor {
+    pub fn new(search_location: Location) -> Self {
+        HoverVisitor {
+            search_location,
+            found_message: None,
+        }
+    }
+
+    fn is_location_in_span(&self, location: &Location, span: &Span) -> bool {
+        let start = &span.start;
+        let end = &span.end;
+
+        location.row >= start.row
+            && location.row <= end.row
+            && location.col >= start.col
+            && location.col <= end.col
+    }
+
+    fn describe_primitive_type(&self, primitive_type: &Primitive) -> Option<String> {
```

Comment:
This function signature should be:
```
fn describe_primitive_type(primitive_type: &Primitive) -> &'static str
```
1) It doesn't use `self` anywhere, so it's not a needed parameter.
2) It never returns `None`. We should _always_ return a message for a primitive.
3) Allocating a new `String` is unnecessary. You're always returning a string literal.
    So returning the `str` directly is more efficient, and makes your logic simpler.